### PR TITLE
Resolve parallel build compilation failure

### PR DESCRIPTION
**Pull request by JackPGreen**
When run with a parallel build (i.e. `mvn -T0.5C`), [some users](https://hazelcast.slack.com/archives/C07066ELRRD/p1716294596589809) are reporting compilation errors:
```
[ERROR] COMPILATION ERROR : 
[INFO] -------------------------------------------------------------
[ERROR] /home/frantisek/work/hz/repos/hazelcast-mono/hazelcast/hazelcast-spring/src/test/java/com/hazelcast/NoMixedJUnitAnnotationsInOurTestSourcesTest.java:[19,35] package com.hazelcast.test.archunit does not exist
```

Further investigation (with `--errors` parameter) shows additional information:
```
org.apache.maven.lifecycle.LifecycleExecutionException: Failed to execute goal org.apache.maven.plugins:maven-compiler-plugin:3.13.0:testCompile (default-testCompile) on project hazelcast-spring: Compilation failure
    at org.apache.maven.lifecycle.internal.MojoExecutor.doExecute2 (MojoExecutor.java:333)
    at org.apache.maven.lifecycle.internal.MojoExecutor.doExecute (MojoExecutor.java:316)
    at org.apache.maven.lifecycle.internal.MojoExecutor.execute (MojoExecutor.java:212)
    at org.apache.maven.lifecycle.internal.MojoExecutor.execute (MojoExecutor.java:174)
    at org.apache.maven.lifecycle.internal.MojoExecutor.access$000 (MojoExecutor.java:75)
    at org.apache.maven.lifecycle.internal.MojoExecutor$1.run (MojoExecutor.java:162)
    at org.apache.maven.plugin.DefaultMojosExecutionStrategy.execute (DefaultMojosExecutionStrategy.java:39)
    at org.apache.maven.lifecycle.internal.MojoExecutor.execute (MojoExecutor.java:159)
    at org.apache.maven.lifecycle.internal.LifecycleModuleBuilder.buildProject (LifecycleModuleBuilder.java:105)
    at org.apache.maven.lifecycle.internal.builder.multithreaded.MultiThreadedBuilder$1.call (MultiThreadedBuilder.java:193)
    at org.apache.maven.lifecycle.internal.builder.multithreaded.MultiThreadedBuilder$1.call (MultiThreadedBuilder.java:180)
    at java.util.concurrent.FutureTask.run (FutureTask.java:317)
    at java.util.concurrent.Executors$RunnableAdapter.call (Executors.java:572)
    at java.util.concurrent.FutureTask.run (FutureTask.java:317)
    at java.util.concurrent.ThreadPoolExecutor.runWorker (ThreadPoolExecutor.java:1144)
    at java.util.concurrent.ThreadPoolExecutor$Worker.run (ThreadPoolExecutor.java:642)
    at java.lang.Thread.run (Thread.java:1583)
Caused by: org.apache.maven.plugin.compiler.CompilationFailureException: Compilation failure
    at org.apache.maven.plugin.compiler.AbstractCompilerMojo.execute (AbstractCompilerMojo.java:1309)
    at org.apache.maven.plugin.compiler.TestCompilerMojo.execute (TestCompilerMojo.java:208)
    at org.apache.maven.plugin.DefaultBuildPluginManager.executeMojo (DefaultBuildPluginManager.java:126)
    at org.apache.maven.lifecycle.internal.MojoExecutor.doExecute2 (MojoExecutor.java:328)
    at org.apache.maven.lifecycle.internal.MojoExecutor.doExecute (MojoExecutor.java:316)
    at org.apache.maven.lifecycle.internal.MojoExecutor.execute (MojoExecutor.java:212)
    at org.apache.maven.lifecycle.internal.MojoExecutor.execute (MojoExecutor.java:174)
    at org.apache.maven.lifecycle.internal.MojoExecutor.access$000 (MojoExecutor.java:75)
    at org.apache.maven.lifecycle.internal.MojoExecutor$1.run (MojoExecutor.java:162)
    at org.apache.maven.plugin.DefaultMojosExecutionStrategy.execute (DefaultMojosExecutionStrategy.java:39)
    at org.apache.maven.lifecycle.internal.MojoExecutor.execute (MojoExecutor.java:159)
    at org.apache.maven.lifecycle.internal.LifecycleModuleBuilder.buildProject (LifecycleModuleBuilder.java:105)
    at org.apache.maven.lifecycle.internal.builder.multithreaded.MultiThreadedBuilder$1.call (MultiThreadedBuilder.java:193)
    at org.apache.maven.lifecycle.internal.builder.multithreaded.MultiThreadedBuilder$1.call (MultiThreadedBuilder.java:180)
    at java.util.concurrent.FutureTask.run (FutureTask.java:317)
    at java.util.concurrent.Executors$RunnableAdapter.call (Executors.java:572)
    at java.util.concurrent.FutureTask.run (FutureTask.java:317)
    at java.util.concurrent.ThreadPoolExecutor.runWorker (ThreadPoolExecutor.java:1144)
    at java.util.concurrent.ThreadPoolExecutor$Worker.run (ThreadPoolExecutor.java:642)
    at java.lang.Thread.run (Thread.java:1583)
```

When [the `maven-compiler-plugin` tries to access the `path` of the artifact](https://github.com/apache/maven-compiler-plugin/blob/a1415aa3833df5985ecdc2b854ef516d3afbf348/src/main/java/org/apache/maven/plugin/compiler/CompilerMojo.java#L313), this is `null` as it's not yet been resolved.

[`null` artifact paths lead me to identify that `ResolutionScope` of the attribution Mojos was incorrect](https://stackoverflow.com/a/21188024), causing Maven to execute these before they've been resolved.

The solution is to broaden the `ResolutionScope` of the `requiresDependencyResolution` as well - looking back at the original report, it's a `test` dependency that's caused the issue.

Tested by updating the reproducer to use a new `SNAPSHOT` version of the `attribution-maven-plugin` and building that locally and re-running, where the error no longer occurs.

### CODE REVIEW

File: src/main/java/com/hazelcast/maven/attribution/AggregateAttributionMojo.java [hunk 1]
```diff
@@ -19,7 +19,7 @@
  */
 @Mojo(name = "aggregate", defaultPhase = LifecyclePhase.PROCESS_RESOURCES,
         aggregator = true, requiresDependencyCollection = ResolutionScope.COMPILE_PLUS_RUNTIME,
-        requiresDependencyResolution = ResolutionScope.NONE,
+        requiresDependencyResolution = ResolutionScope.TEST,
```
Comment (ldziedziul) [inline src/main/java/com/hazelcast/maven/attribution/AggregateAttributionMojo.java:22]: Why `TEST` and not `COMPILE_PLUS_RUNTIME` ?

Comment (JackPGreen) [reply src/main/java/com/hazelcast/maven/attribution/AggregateAttributionMojo.java:22]: > Why `TEST` and not `COMPILE_PLUS_RUNTIME` ?

Short answer - it doesn't work if you don't.

Longer answer - it's failing with compilation errors about being unable to access `test` scoped artefacts.

**Review by kwart — APPROVED** [review]
It seems the change does the trick :tada: 
I also tried to play with the `requiresDependencyResolution` value, but didn't try the `TEST` sope :facepalm: 

Thanks for the fix, Jack.

**Review by ldziedziul — APPROVED** [review]
Nice work 💪

**Pull request merged**